codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketCloseStatus.java

```diff
@@ -0,0 +1,236 @@
+    public static final WebSocketCloseStatus INVALID_MESSAGE_TYPE = valueOf(1003, "Invalid message type");
+    public static final WebSocketCloseStatus INVALID_PAYLOAD_DATA = valueOf(1007, "Invalid payload data");
+    public static final WebSocketCloseStatus POLICY_VIOLATION = valueOf(1008, "Policy violation");
+    public static final WebSocketCloseStatus MESSAGE_TOO_BIG = valueOf(1009, "Message too big");
+    public static final WebSocketCloseStatus MANDATORY_EXTENSION = valueOf(1010, "Mandatory extension");
+    public static final WebSocketCloseStatus INTERNAL_SERVER_ERROR = valueOf(1011, "Internal server error");
+
+    // 1004, 1005, 1006, 1015 are reserved and should never be used by user
+    //public static final WebSocketCloseStatus SPECIFIC_MEANING = valueOf(1004, "...");
+    //public static final WebSocketCloseStatus EMPTY = valueOf(1005, "Empty");
+    //public static final WebSocketCloseStatus ABNORMAL_CLOSURE = valueOf(1006, "Abnormal closure");
+    //public static final WebSocketCloseStatus TLS_HANDSHAKE_FAILED(1015, "TLS handshake failed");
+
+    private final int statusCode;
+    private volatile String reasonText;
+
+    private WebSocketCloseStatus(int statusCode, String reasonText) {
+        if (!isValidRfc6455Code(statusCode)) {
+            throw new IllegalArgumentException(
+                "WebSocket close status code does NOT comply with RFC-6455: " + statusCode);
+        }
+        this.statusCode = statusCode;
+        this.reasonText = reasonText;
+    }
+
+    public int code() {
+        return statusCode;
+    }
+
+    public String reasonText() {
+        return reasonText == null ? "Close status #" + statusCode : reasonText;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (null == o || getClass() != o.getClass()) {
+            return false;
+        }
+
+        WebSocketCloseStatus that = (WebSocketCloseStatus) o;
+
+        return statusCode == that.statusCode;
+    }
+
+    @Override
+    public int hashCode() {
+        return statusCode;
+    }
+
+    @Override
+    public String toString() {
+        // E.g.: "1000 Bye", "1009 Message too big"
+        return code() + " " + reasonText();
+    }
+
+    public static boolean isValidRfc6455Code(int code) {
+        return code < 0 ||
+            1000 <= code && code <= 1003 ||
+            1007 <= code && code <= 1011 ||
+            3000 <= code;
+    }
+
+    public static WebSocketCloseStatus valueOf(int code) {
+        return valueOf(code, null);
+    }
+
+    public static WebSocketCloseStatus valueOf(int code, String message) {
+        WebSocketCloseStatus status = VALUES.get(code);
+        if (status == null) {
+            VALUES.putIfAbsent(code, new WebSocketCloseStatus(code, message));
+            status = VALUES.get(code);
+        }
+        if (message != null) {
+            status.reasonText = message;
```

Comment:
This is racy and may even override the reasonText if it was already in the map.

Reply:
exactly what I need. I want to override reasonText applications-wide.
in 99.9% cases nobody will use this feature and close statuses will be registered once per app start. however in the last 0.1% it would be impossible to override default 'reasonText' without such approach. See 'testUserLaterDefinedStatuses' test as an example.

Reply:
re-worked and simplified